site/gatsby-site/src/utils/classy.js

```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+
+const classy = (type, classStringOrFunction) => {
+  const component = (props) => {
+
+    const componentClassName = (
+      typeof classStringOrFunction == 'function'
+        ? classStringOrFunction(props)
+        : classStringOrFunction
+    );
+
+    const className = (
+      [componentClassName, props.className]
+        .filter((e) => e)
+        .join(' ') 
+    ) || undefined;
+
+    return React.createElement(
+      type,
+      { ...props, className },
+      Array.isArray(props.children) ? props.children : [props.children]
+    );
+  }
+
+  component.displayName = 'Classy' + type.slice(0, 1).toUpperCase() + type.slice(1);
+
+  return component;
+};
+
+const classyDiv = (className) => classy('div', className);
+
+const classySpan = (className) => classy('span', className);
+
+export { classy, classyDiv, classySpan };
```

Comment:
I'm fine with changing this for consistency, but for the record I think this approach is the superior one.

Let's take something likes this:

``` javascript
  return (
    <RiskDetails open={risk.startClosed ? undefined : true} generated={risk.generated}>
      {/* ... */}
    </RiskDetails>
  );
}

{/* ... */}

const RiskDetails = classy(
  'details',
  ({ generated }) => `
  ${
    generated ? 'border-gray-400' : 'border-red-700'
  } border-l-2 border-r-2 border-t-2 open:border-b-2
  h-0 open:h-fit
  relative max-w-full
  open:p-3 md:open:p-6 open:rounded
  cursor-pointer

  [&[open]>summary]:before:content-['⏷']
        [&>summary]:before:content-['⏵']
  [&[open]>summary]:before:w-4
        [&>summary]:before:w-4
`
);
```

How would we rewrite it? One way is to just shove everything in one line.

``` javascript
  return (
    <details open={risk.startClosed ? undefined : true} className={` ${ risk.generated ? 'border-gray-400' : 'border-red-700' } border-l-2 border-r-2 border-t-2 open:border-b-2 h-0 open:h-fit relative max-w-full open:p-3 md:open:p-6 open:rounded cursor-pointer [&[open]>summary]:before:content-['⏷'] [&>summary]:before:content-['⏵'] [&[open]>summary]:before:w-4 [&>summary]:before:w-4`}>
      {/* ... */}
    </details>
  );
}
```

This makes it hard to visually pick out groups of related classes, doesn't give us a name saying what this `<details>` element is for, and produces worse diffs.

We could keep it inline but use a multi-line string.

``` javascript
  return (
    <details open={risk.startClosed ? undefined : true} className={`
      ${
        risk.generated ? 'border-gray-400' : 'border-red-700'
      } border-l-2 border-r-2 border-t-2 open:border-b-2
      h-0 open:h-fit
      relative max-w-full
      open:p-3 md:open:p-6 open:rounded
      cursor-pointer

      [&[open]>summary]:before:content-['⏷']
            [&>summary]:before:content-['⏵']
      [&[open]>summary]:before:w-4
            [&>summary]:before:w-4
    `}>
      {/* ... */}
    </details>
  );
}
```

This makes the style information easier to read, but it makes the markup harder to read because it violates the expectation that an indented block below `<details` will be the element body, when in fact it's just an extremely long attribute value.

We could move the styling elsewhere by declaring a variable or or a function to generated it if it's dynamic.

``` javascript
  return (
    <details className={riskDetails({ generated: risk.generated})} open={risk.startClosed ? undefined : true}>
      {/* ... */}
    </details>
  );
}

const riskDetails = ({ generated }) => `
  ${
    risk.generated ? 'border-gray-400' : 'border-red-700'
  } border-l-2 border-r-2 border-t-2 open:border-b-2
  h-0 open:h-fit
  relative max-w-full
  open:p-3 md:open:p-6 open:rounded
  cursor-pointer

  [&[open]>summary]:before:content-['⏷']
        [&>summary]:before:content-['⏵']
  [&[open]>summary]:before:w-4
        [&>summary]:before:w-4
`;
```

This is ok -- `riskDetails` is even sort of a name, but it's only on the opening tag, which can make it harder to find the matching closing tag, especially if there's a lot between them and it's a common tag like `<div>`.

We could create a component that just wraps its children in a styled element:

``` javascript
  return (
    <RiskDetails open={risk.startClosed ? undefined : true} generated={risk.generated}>
      {/* ... */}
    </RiskDetails>
  );
}

{/* ... */}

const RiskDetails = (props) => (
  <details {...{
    ...props, 
    className: `
      ${
        props.generated ? 'border-gray-400' : 'border-red-700'
      } border-l-2 border-r-2 border-t-2 open:border-b-2
      h-0 open:h-fit
      relative max-w-full
      open:p-3 md:open:p-6 open:rounded
      cursor-pointer

      [&[open]>summary]:before:content-['⏷']
            [&>summary]:before:content-['⏵']
      [&[open]>summary]:before:w-4
            [&>summary]:before:w-4
      ${props.className}
    `
  }}>
    {props.children}
  </details>
);
```

That's more or less fine, but if you do this many times, the boilerplate is annoying. We have to correctly deal with `children`, `className`, and `props` every time.

``` javascript
const ComponentOne = (props) => (
  <div {...{
    ...props, 
    className: `
      ${/* Style */}
      ${props.className}
    `
  }}>
    {props.children}
  </div>
);
const ComponentTwo = (props) => (
  <div {...{
    ...props, 
    className: `
      ${/* Style */}
      ${props.className}
    `
  }}>
    {props.children}
  </div>
);
```

There are only really three things that matter for this kind of component: The component name, the className, and the base element type. So I wrote a function that takes just those three things.

``` javascript
const ComponentOne = classy('div', /* Style */);
const ComponentTwo = classy('div', /* Style */);
```

I don't see how that's worse. The arguments I can see are

-   **We have to maintain the extra 30 lines of code this adds.** -- but this is trivial given the size of our codebase.

-   **It makes it more cumbersome to change from a style-only component to a full-fledged component** -- true, but it makes less cumbersome to change from an inline style to a named component, so it evens out.

-   **It's less transparent than writing a component normally** -- I think the basic form where we just provide a string with a list of classes is fairly transparent.

    ``` javascript
    const ComponentOne = classyDiv('p-1 m-1 rounded bg-gray-300')
    ```

    It's slightly less obvious when we use a function

    ``` javascript
    const ComponentOne = classyDiv(
      ({ primary }) => `p-1 m-1 rounded ${primary ? 'bg-blue-300' : 'bg-gray-300'}`
    );
    ```

    But still, this is exactly the same way its done in Styled Components, and I didn't find that hard to understand.

-   **Naming things and separating the style from the markup defeats the purpose of using Tailwind** -- Tailwind makes it so that you *can* leave things unnamed and include styles with media queries, etc. inline with the markup. But that doesn't mean that we *should always* do so. Sometimes it makes sense to separate style from markup and sometimes it doesn't.

Reply:
I got your points, and after thinking about these different approaches, and the consistency of the code, I think we should follow the Component approach you mentioned above instead of using the `classy` solution. It's a solution between complete inline styling and naming components solution.
I mean this:

```
const ComponentOne = (props) => (
  <div {...{
    ...props, 
    className: `
      ${/* Style */}
      ${props.className}
    `
  }}>
    {props.children}
  </div>
);
const ComponentTwo = (props) => (
  <div {...{
    ...props, 
    className: `
      ${/* Style */}
      ${props.className}
    `
  }}>
    {props.children}
  </div>
);
```
Could you please change it?

Reply:
Done!